### ISSUE TITLE
Readiness probe for pgbouncer

### ISSUE THREAD

**Issue by zerkms** [CrunchyData/postgres-operator#3056]
Have an idea to improve PGO? We'd love to hear it! We're going to need some information from you to learn more about your feature requests.

Please be sure you've done the following:

- [ ] Provide a concise description of your feature request.
- [ ] Describe your use case. Detail the problem you are trying to solve.
- [ ] Describe how you envision that the feature would work.
- [ ] Provide general information about your current PGO environment.

## Overview

At the moment it's possible to try to connect to a pgbouncer while it's still not ready to accept connections. It happens because kubernetes marks it as in `Ready` state as long as the container starts, but it still takes some time for the application to bind to the port and start accepting connections.

## Use Case

One of the aims of pgbouncer is to increase reliability/availability of a postgresql cluster. But without readiness probe there is a little window of opportunity to try to connect to a pgbouncer instance that is not ready to serve connections, so the application would get a connection error.

## Desired Behavior

I'm not sure if it should be a script-based readiness check that ensures the pgbouncer is up and running fine, or if just a tcp port check would suffice.

## Environment

Tell us about your environment:

Please provide the following details:

- Platform: (`Kubernetes`, `OpenShift`, `Rancher`, `GKE`, `EKS`, `AKS` etc.)
- Platform Version: (e.g. `1.20.3`, `4.7.0`)
- PGO Image Tag: (e.g. `ubi8-5.1.0-0`)
- Postgres Version (e.g. `14`)
- Storage: (e.g. `hostpath`, `nfs`, or the name of your storage class)
- Number of Postgres clusters: (`XYZ`)

## Additional Information

Please provide any additional information that may be helpful.

**Comment by tjmoore4** [comment]
Hello @zerkms. Please provide a bit more information about your environment, particularly your PGO version, platform and version, and how you are connecting to pgBouncer. This will help us to better understand your desired use-case.

**Comment by zerkms** [comment]
Oh right, sorry

> PGO version

5.0.4

> platform

baremetal, kubeadm, 1.22.6 (Major - 1)

> how you are connecting to pgBouncer

Using `db-pgbouncer` service which is created and managed by the operator.

The use case: to connect to pgbouncer that is _ready_ to accept connections. As it is now (without readiness probe) it's possible that one tries to establish a connection with a pod that is not ready to accept connections yet.

**Comment by ValClarkson** [comment]
HI @zerkms, Can you give us an example where you are encountering errors saying pgbouncer is not available with some logs?  We are trying to understand your use case a bit better.

**Comment by zerkms** [comment]
@ValClarkson what exactly do you expect to see? It's a generic postgresql server connectivity issue.

Please check the pgbouncer manifest that the operator (v5) generates and explain the order of these 2 events:

1. Container/pod is marked as `Ready`
2. The pgbouncer process fully started (bound to a socket and is accepting new connection)

Should it be 1 then 2, or 2 then 1? Who guarantees it?

**Comment by cbandy** [comment]
> At the moment it's possible to try to connect to a pgbouncer while it's still not ready to accept connections. It happens because kubernetes marks it as in `Ready` state as long as the container starts, but it still takes some time for the application to bind to the port and start accepting connections.

How much time are you seeing between PgBouncer start and it binding to the port?

I see it taking between 5-30ms on a variety of systems which is consistently faster than the Endpoint controller can respond.

<br/>

> But without readiness probe there is a little window of opportunity to try to connect to a pgbouncer instance that is not ready to serve connections, so the application would get a connection error.
>
> > how you are connecting to pgBouncer
> 
> Using `db-pgbouncer` service which is created and managed by the operator.

Excellent. PgBouncer is ready and accepting connections when this Service routes connections to it.

<br/>

> what exactly do you expect to see? It's a generic postgresql server connectivity issue.
>
> Please check the pgbouncer manifest that the operator (v5) generates and explain the order of these 2 events:
> 
> 1. Container/pod is marked as `Ready`
> 2. The pgbouncer process fully started (bound to a socket and is able to accept tcp connections)
> 
> Should it be 1 then 2, or 2 then 1? Who guarantees it?

Without a readiness probe, the order is not guaranteed. Kubelet starts the process then notifies the API server that the pod is Ready. Of course, information being in the API is not the same as network connectivity.

It sounds to me like you have not experienced any connectivity issues. If you do, please reopen this with details. Thanks!

**Comment by zerkms** [comment]
> How much time are you seeing between PgBouncer start and it binding to the port?

enough to experience the problem several times.

> Without a readiness probe, the order is not guaranteed. Kubelet starts the process then notifies the API server that the pod is Ready. Of course, information being in the API is not the same as network connectivity.

That's correct: the order is not guaranteed, hence it's possible that a service traffic is routed to a pod that is NOT READY to accept connections, hence the client will see an error.

> Excellent. PgBouncer is ready and accepting connections when this Service routes connections to it.

That's not true.

> It sounds to me like you have not experienced any connectivity issues.

Every network service *must* have readiness probe.

> If you do, please reopen this with details. Thanks!

What kind of details you need, I seriously don't understand: the client tries to connect, but the socket is not listened yet, hence a client receives a standard tcp connection error.

**Comment by zerkms** [comment]
I have re-read my message, it sounds rude. Please count it as a language barrier, and my apologies.

**Comment by zerkms** [comment]
> Does the PgBouncer Pod exist, for example? Please describe the steps you took and more details about the error message.

My scenario was the following:

1. There is a python script that periodically (every 10 seconds) connects to a database via a pgbouncer service
2. There are 2 pgbouncer pods

At one point I drained the node, so a pgbouncer pod was deleted on that node and rescheduled on another node.

Then I have noticed a tcp connection error from the python script.

The timestamp (with milliseconds) of the failure matched the very moment when the pgbouncer started (using the pgbouncer logs).

I agree it does not prove anything, but I don't believe in coincidences: to me what happened is - pgbouncer service got marked as Ready before pgbouncer process was ready to accept network connections.
